src/storage/archive_index.rs

```diff
@@ -0,0 +1,114 @@
+use crate::error::Result;
+use crate::storage::{compression::CompressionAlgorithm, FileRange};
+use failure::ResultExt;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::io;
+use std::path::{Path, PathBuf};
+
+#[derive(Deserialize, Serialize)]
+pub struct FileInfo {
+    range: FileRange,
+    compression: CompressionAlgorithm,
+}
+
+impl FileInfo {
+    pub fn range(&self) -> FileRange {
+        self.range.clone()
+    }
+    pub fn compression(&self) -> CompressionAlgorithm {
+        self.compression
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct Index {
+    files: HashMap<PathBuf, FileInfo>,
+}
+
+impl Index {
+    pub fn load(reader: impl io::Read) -> Result<Index> {
+        serde_cbor::from_reader(reader)
+            .context("deserialization error")
+            .map_err(Into::into)
+    }
+
+    pub fn save(&self, writer: impl io::Write) -> Result<()> {
+        serde_cbor::to_writer(writer, self)
+            .context("serialization error")
+            .map_err(Into::into)
+    }
+
+    pub fn new_from_zip<R: io::Read + io::Seek>(zipfile: &mut R) -> Result<Index> {
+        let mut archive = zip::ZipArchive::new(zipfile)?;
+
+        // get file locations
+        let mut files: HashMap<PathBuf, FileInfo> = HashMap::with_capacity(archive.len());
+        for i in 0..archive.len() {
+            let zf = archive.by_index(i)?;
+
+            files.insert(
+                PathBuf::from(zf.name()),
+                FileInfo {
+                    range: FileRange::new(
+                        zf.data_start(),
+                        zf.data_start() + zf.compressed_size() - 1,
+                    ),
+                    compression: match zf.compression() {
+                        zip::CompressionMethod::Bzip2 => CompressionAlgorithm::Bzip2,
+                        c => failure::bail!("unsupported compression algorithm {} in zip-file", c),
```

Comment:
Hmm, technically we could probably support `Stored` (no compression), but since we don't use that in practice this seems fine.

Reply:
👍